test/expect-tests/dune_pkg/dune_pkg_unit_tests.ml

```diff
@@ -27,12 +28,40 @@ let url ~port ~filename =
   let localhost = Unix.inet_addr_loopback |> Unix.string_of_inet_addr in
   Format.sprintf "http://%s:%d/%s" localhost port filename |> OpamUrl.of_string
 
-let download ~port ~filename ~target () =
+let calculate_checksum ~filename =
+  OpamHash.compute filename |> Checksum.of_opam_hash
+
+let wrong_checksum =
+  OpamHash.compute_from_string "random content" |> Checksum.of_opam_hash
+
+let download ~port ~filename ~target ?checksum () =
   let open Fiber.O in
   let url = url ~port ~filename in
-  let* () = Fetch.fetch Loc.none url ~target in
-  print_endline "Done downloading";
-  Fiber.return ()
+  let* res = Fetch.fetch ~checksum ~target url in
+  match res with
+  | Error (Unavailable (_url, normal)) ->
+    let errs = [ Pp.text "Failure while downloading" ] in
+    let errs =
+      match normal with
+      | None -> errs
+      | Some s -> Pp.textf "Opam says: %S" s :: errs
+    in
+    User_error.raise ~loc:Loc.none errs
+  | Error (Checksum_mismatch actual_checksum) ->
+    let expected_checksum =
+      match checksum with
+      | Some v -> v
+      | None -> assert false
+    in
+    User_error.raise ~loc:Loc.none
```

Comment:
You should probably take the location of the checksum as an argument. In fact, the type of `?checksum` should be `(Loc.t * Checksum.t) option`